server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java

```diff
@@ -0,0 +1,358 @@
+                    .withRole(ShardRouting.Role.DEFAULT)
+                    .build();
+                indexShardRoutingBuilder.addShard(replicaShardRouting);
+                nodeToShardRoutings.putIfAbsent(searchNode, new ArrayList<>());
+                nodeToShardRoutings.get(searchNode).add(replicaShardRouting);
+            }
+            indexRoutingTableBuilder.addIndexShard(indexShardRoutingBuilder);
+            routingTableBuilder.add(indexRoutingTableBuilder.build());
+        }
+
+        if (allowMaster) {
+            ShardId lastPrimaryShardId = new ShardId(indexMetadata.getIndex(), numberOfPrimaryShards - 1);
+            masterPrimaryShardRouting = TestShardRouting.newShardRouting(
+                lastPrimaryShardId,
+                masterNode.getId(),
+                null,
+                true,
+                ShardRoutingState.STARTED
+            );
+            IndexShardRoutingTable.Builder indexShardRoutingBuilderMasterNode = IndexShardRoutingTable.builder(lastPrimaryShardId);
+            indexShardRoutingBuilderMasterNode.addShard(masterPrimaryShardRouting);
+            indexRoutingTableBuilder.addIndexShard(indexShardRoutingBuilderMasterNode);
+            routingTableBuilder.add(indexRoutingTableBuilder.build());
+        }
+
+        metadataBuilder.put(projectBuilder).generateClusterUuidIfNeeded();
+        state.nodes(discoveryNodeBuilder);
+        state.metadata(metadataBuilder);
+        state.routingTable(GlobalRoutingTable.builder().put(projectId, routingTableBuilder).build());
+        clusterState = state.build();
+
+        routingIndexNodeOne = RoutingNodesHelper.routingNode(
+            indexNodeOne.getId(),
+            indexNodeOne,
+            nodeToShardRoutings.get(indexNodeOne).toArray(new ShardRouting[0])
+        );
+        routingIndexNodeTwo = RoutingNodesHelper.routingNode(
+            indexNodeTwo.getId(),
+            indexNodeTwo,
+            nodeToShardRoutings.get(indexNodeTwo).toArray(new ShardRouting[0])
+        );
+        routingSearchNodeOne = RoutingNodesHelper.routingNode(
+            searchNodeOne.getId(),
+            searchNodeOne,
+            nodeToShardRoutings.get(searchNodeOne).toArray(new ShardRouting[0])
+        );
+        routingSearchNodeTwo = RoutingNodesHelper.routingNode(
+            searchNodeTwo.getId(),
+            searchNodeTwo,
+            nodeToShardRoutings.get(searchNodeTwo).toArray(new ShardRouting[0])
+        );
+        if (allowMaster) {
+            routingMasterNode = RoutingNodesHelper.routingNode(masterNode.getId(), masterNode, masterPrimaryShardRouting);
+        } else {
+            routingMasterNode = RoutingNodesHelper.routingNode(masterNode.getId(), masterNode);
+        }
+    }
+
+    public void testCanAllocate_under_threshold() {
+        setup(true);
+
+        Settings settings = Settings.builder()
```

Comment:
L195-L215 is duplicate code across tests. Since you have the `setup` method, can you deduplicate the test code?